src/pkg/systemartifact/manager.go

```diff
@@ -0,0 +1,231 @@
+	err = mgr.regCli.PushBlob(repoName, artifactRecord.Digest, artifactRecord.Size, reader)
+
+	if err != nil {
+
+		mgr.dao.Delete(ctx, artifactRecord.Vendor, artifactRecord.Repository, artifactRecord.Digest)
+		return int64(0), err
+	}
+	return id, nil
+}
+
+func (mgr *systemArtifactManager) Read(ctx context.Context, vendor string, repository string, digest string) (io.ReadCloser, error) {
+	sa, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return nil, err
+	}
+	repoName := mgr.getRepositoryName(vendor, repository)
+	_, readCloser, err := mgr.regCli.PullBlob(repoName, sa.Digest)
+	if err != nil {
+		return nil, err
+	}
+	return readCloser, nil
+}
+
+func (mgr *systemArtifactManager) Delete(ctx context.Context, vendor string, repository string, digest string) error {
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	err := mgr.regCli.DeleteBlob(repoName, digest)
+
+	if err != nil {
+		return err
+	}
+
+	err = mgr.dao.Delete(ctx, vendor, repository, digest)
+
+	return err
+}
+
+func (mgr *systemArtifactManager) Exists(ctx context.Context, vendor string, repository string, digest string) (bool, error) {
+	_, err := mgr.dao.Get(ctx, vendor, repository, digest)
+	if err != nil {
+		return false, err
+	}
+
+	repoName := mgr.getRepositoryName(vendor, repository)
+	exist, err := mgr.regCli.BlobExist(repoName, digest)
+
+	if err != nil {
+		return false, err
+	}
+
+	return exist, nil
+}
+
+func (mgr *systemArtifactManager) GetStorageSize(ctx context.Context) (int64, error) {
+	listQuery := q.Query{}
+	artifacts, err := mgr.dao.List(ctx, &listQuery)
+
+	if err != nil {
+		return int64(0), err
+	}
+
+	size := int64(0)
+
+	for _, artifact := range artifacts {
+		size += artifact.Size
+	}
+	return size, nil
+}
+
+func (mgr *systemArtifactManager) RegisterCleanupCriteria(vendor string, artifactType string, criteria CleanupCriteria) {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	cleanupCriterias[key] = criteria
+}
+
+func (mgr *systemArtifactManager) GetCleanupCriteria(vendor string, artifactType string) CleanupCriteria {
+	key := fmt.Sprintf(keyFormat, vendor, artifactType)
+	if criteria, ok := cleanupCriterias[key]; ok {
+		return criteria
+	}
+	return DefaultCleanupCriteria
+}
+
+func (mgr *systemArtifactManager) Cleanup(ctx context.Context) (int64, int64, error) {
+	logger.Info("Starting system artifact cleanup")
+	// clean up artifact records having customized cleanup criteria first
+	totalReclaimedSize := int64(0)
+	totalRecordsDeleted := int64(0)
+
+	for key, val := range cleanupCriterias {
+		logger.Infof("Executing cleanup for 'vendor:artifactType' : %s", key)
+		deleted, size, err := mgr.cleanup(ctx, val)
+		totalRecordsDeleted += deleted
+		totalReclaimedSize += size
+
+		if err != nil {
+			logger.Errorf("Error when cleaning up system artifacts for 'vendor:artifactType':%s, %v", key, err)
+			return totalRecordsDeleted, totalReclaimedSize, err
```

Comment:
For some storage types, it might cause an error when deleting a file, should we log the error and continue the cleanup operation? because it fails, then it will always fail on the same error,  the system artifact might be accumulated.